src/main/java/no/digipost/collection/NonEmptyList.java

```diff
@@ -0,0 +1,239 @@
+    }
+
+
+    /**
+     * Try to construct a non-empty list from a given list, which may be empty.
+     *
+     * @param <E> the type of elements in the list.
+     * @param list the list, which may be empty
+     *
+     * @return the resulting non-empty list, or {@link Optional#empty()} if the
+     *         given list is empty
+     */
+    static <E> Optional<NonEmptyList<E>> of(List<E> list) {
+        return firstOf(list).map(first -> NonEmptyList.of(first, list.subList(1, list.size())));
+    }
+
+
+    /**
+     * <strong>Unsafe</strong> construction of non-empty list from a possibly empty list.
+     * <p>
+     * This method should only be used when the given list is <em>guarantied</em>
+     * to be empty, and thus offers a fail-fast way to introduce the non-empty
+     * quality on a type level. Use {@link #copyOf(List)} if you need
+     * more flexibility in handling of a possible empty list.
+     *
+     * @param <E> the type of elements in the list.
+     * @param list the list
+     *
+     * @return the resulting non-empty list
+     *
+     * @throws IllegalArgumentException if the given list is empty
+     *
+     * @see #copyOf(List)
+     */
+    static <E> NonEmptyList<E> ofUnsafe(List<E> list) {
+        return copyOf(list).orElseThrow(() -> new IllegalArgumentException("empty list"));
+    }
+
+
+    /**
+     * Try to construct a non-empty list from copying the elements
+     * of a given list, which may be empty.
+     *
+     * @param <E> the type of elements in the list.
+     * @param list the list, which may be empty
+     *
+     * @return the resulting non-empty list,
+     *         or {@link Optional#empty()} if the given list is empty
+     */
+    static <E> Optional<NonEmptyList<E>> copyOf(List<E> list) {
+        return firstOf(list).map(first -> NonEmptyList.of(first, new ArrayList<>(list.subList(1, list.size()))));
+    }
+
+
+    /**
+     * <strong>Unsafe</strong> construction of non-empty list from copying the
+     * elements of a possibly empty list.
+     * <p>
+     * This method should only be used when the given list is <em>guarantied</em>
+     * to be empty, and thus offers a fail-fast way to introduce the non-empty
+     * quality on a type level. Use {@link #copyOf(List)} if you need
+     * more flexibility in handling of a possible empty list.
+     *
+     * @param <E> the type of elements in the list.
+     * @param nonEmptyList the list
+     *
+     * @return the resulting non-empty list
+     *
+     * @throws IllegalArgumentException if the given list is empty
+     *
+     * @see #copyOf(List)
+     */
+    static <E> NonEmptyList<E> copyOfUnsafe(List<E> nonEmptyList) {
+        return copyOf(nonEmptyList).orElseThrow(() -> new IllegalArgumentException("empty list"));
+    }
+
+
+    static <T> Collector<T, ?, NonEmptyList<T>> toNonEmptyListOrIfEmptyThrow(Supplier<? extends RuntimeException> exceptionSupplier) {
```

Comment:
```suggestion
    static <T> Collector<T, ?, NonEmptyList<T>> toNonEmptyListOrElseThrow(Supplier<? extends RuntimeException> exceptionSupplier) {
```

Reply:
Minner kanskje mer om navnene brukt for f.eks `Optional`?

Reply:
Why do you prefer "OrElse" instead of "OrIfEmpty"? I used "OrIfEmpty" to communicate that it will throw if it (the stream acted on) is empty, and it may not be that clear what "else" implicitly means. Then it just says "collect to non-empty list. Or not". Saying it will throw if the source/stream is empty also helps with what an exception should look like, which one must supply.